### PR TITLE
vmu: Add VMU framebuffer API

### PR DESCRIPTION
**Pull request by pcercuei**
This API can be used to render and present dynamic images to the VMUs connected to the system. It also supports rendering text using a user-provided font.

### CODE REVIEW

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
+/* KallistiOS ##version##
+
+   dc/vmu_fb.h
+   Copyright (C) 2023 Paul Cercueil
+
+*/
+
+#include <kos.h>
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:8]: Please do not include <kos.h> in any public header files. Include the appropriate header files that you need, please.

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+/** \brief  VMU framebuffer.
+ *
+ *  This object can be used with the API below to compose a 48x32 image that can
+ *  then be displayed on the VMUs connected to the system.
+ */
+typedef struct {
+	uint32 data[48];
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:16]:
```suggestion
    uint32_t data[48];
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+} vmufb_t;
+
+/** \brief  VMU framebuffer font meta-data. */
+typedef struct {
+	unsigned int w;         /**< \brief Character width in pixels */
+	unsigned int h;         /**< \brief Character height in pixels */
+	unsigned int stride;    /**< \brief Size of one charater in bytes */
+	const char *fontdata;   /**< \brief Pointer to the font data */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:24]:
```suggestion
    unsigned int w;         /**< \brief Character width in pixels */
    unsigned int h;         /**< \brief Character height in pixels */
    unsigned int stride;    /**< \brief Size of one character in bytes */
    const char *fontdata;   /**< \brief Pointer to the font data */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ *                          drawing area, in pixels
+ *  \param  w               The width of the drawing area, in pixels
+ *  \param  h               The height of the drawing area, in pixels
+ *  \param  data            A pointer to the pixel data that will be painted
+ *                          into the drawing area.
+ */
+void vmufb_paint_area(vmufb_t *fb,
+		      unsigned int x, unsigned int y,
+		      unsigned int w, unsigned int h,
+		      const char *data);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:45]:
```suggestion
void vmufb_paint_area(vmufb_t *fb,
                      unsigned int x, unsigned int y,
                      unsigned int w, unsigned int h,
                      const char *data);
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ *  \param  x               The horizontal position of the top-left corner of
+ *                          the drawing area, in pixels
+ *  \param  y               The vertical position of the top-left corner of the
+ *                          drawing area, in pixels
+ *  \param  w               The width of the drawing area, in pixels
+ *  \param  h               The height of the drawing area, in pixels
+ */
+void vmufb_clear_area(vmufb_t *fb,
+		      unsigned int x, unsigned int y,
+		      unsigned int w, unsigned int h);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:62]:
```suggestion
void vmufb_clear_area(vmufb_t *fb,
                      unsigned int x, unsigned int y,
                      unsigned int w, unsigned int h);
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+ *  \param  line_spacing    Specify the number of empty lines that should
+ *                          separate two lines of text
+ *  \param  str             The text to render
+ */
+void vmufb_print_string_into(vmufb_t *fb,
+			     const vmufb_font_t *font,
+			     unsigned int x, unsigned int y,
+			     unsigned int w, unsigned int h,
+			     unsigned int line_spacing,
+			     const char *str);
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:105]:
```suggestion
void vmufb_print_string_into(vmufb_t *fb,
                             const vmufb_font_t *font,
                             unsigned int x, unsigned int y,
                             unsigned int w, unsigned int h,
			     unsigned int line_spacing,
			     const char *str);
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ *  \param  font            A pointer to the vmufb_font_t that will be used for
+ *                          painting the text
+ *  \param  str             The text to render
+ */
+static __inline__ void vmufb_print_string(vmufb_t *fb,
+					  const vmufb_font_t *font,
+					  const char *str)
+{
+	vmufb_print_string_into(fb, font, 0, 0, 48, 32, 0, str);
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:122]:
```suggestion
static __inline__ void vmufb_print_string(vmufb_t *fb,
                                          const vmufb_font_t *font,
                                          const char *str) {
    vmufb_print_string_into(fb, font, 0, 0, 48, 32, 0, str);
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 8]
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:1]: Please provide a correct header comment here.

Also, please remember that KOS' core kernel code is not GPL. I will not be able to accept this PR for the kernel if you intend for this to be GPL licensed.

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <dc/maple/vmu.h>
+#include <dc/vmu_fb.h>
+
+#define GENMASK(h, l) \
+	(((unsigned int)-1 << (l)) & ((unsigned int)-1 >> (31 - (h))))
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:9]:
```suggestion
#define GENMASK(h, l) \
    (((unsigned int)-1 << (l)) & ((unsigned int)-1 >> (31 - (h))))
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+		tmp &= GENMASK(7 - (offt & 0x7), lsb);
+
+		bits |= tmp >> lsb;
+
+		offt += nb_bits;
+		w -= nb_bits;
+	}
+
+	return bits;
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:37]:
```suggestion
static uint64_t extract_bits(const uint8_t *data,
                             unsigned int offt, unsigned int w) {
    uint32_t tmp, lsb, nb_bits;
    uint64_t bits = 0;

    while(w) {
        tmp = data[offt / 8];

        if(8 - (offt & 0x7) > w)
            lsb = 8 - (offt & 0x7) - w;
        else
            lsb = 0;

        nb_bits = 8 - (offt & 0x7) - lsb;
        bits <<= nb_bits;

        tmp &= GENMASK(7 - (offt & 0x7), lsb);

        bits |= tmp >> lsb;

        offt += nb_bits;
        w -= nb_bits;
    }

    return bits;
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+		tmp &= ~mask;
+
+		tmp |= ((bits >> (w - nb_bits)) << lsb) & mask;
+
+		data[offt / 8] = tmp;
+
+		offt += nb_bits;
+		w -= nb_bits;
+	}
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:63]:
```suggestion
static void insert_bits(uint8_t *data, unsigned int offt,
                        unsigned int w, uint64_t bits) {
    uint32_t tmp, lsb, nb_bits, mask;

    while(w) {
        tmp = data[offt / 8];

        if(8 - (offt & 0x7) > w)
            lsb = 8 - (offt & 0x7) - w;
        else
            lsb = 0;

        nb_bits = 8 - (offt & 0x7) - lsb;
        mask = GENMASK(7 - (offt & 0x7), lsb);
        tmp &= ~mask;

        tmp |= ((bits >> (w - nb_bits)) << lsb) & mask;

        data[offt / 8] = tmp;

        offt += nb_bits;
        w -= nb_bits;
    }
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+		      const char *data)
+{
+	unsigned int i;
+	uint64_t bits;
+
+	for (i = 0; i < h; i++) {
+		bits = extract_bits((const uint8_t *)data, i * w, w);
+		insert_bits((uint8_t *)fb->data, (y + i) * 48 + x, w, bits);
+	}
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:77]:
```suggestion
void vmufb_paint_area(vmufb_t *fb,
                      unsigned int x, unsigned int y,
                      unsigned int w, unsigned int h,
                      const char *data) {
    unsigned int i;
    uint64_t bits;

    for(i = 0; i < h; i++) {
        bits = extract_bits((const uint8_t *)data, i * w, w);
        insert_bits((uint8_t *)fb->data, (y + i) * 48 + x, w, bits);
    }
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+{
+	uint32_t tmp[48] = {};
+
+	vmufb_paint_area(fb, x, y, w, h, (const char *) tmp);
+}
+
+void vmufb_present(const vmufb_t *fb, maple_device_t *dev)
+{
+	vmu_draw_lcd_rotated(dev, fb->data);
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:96]:
```suggestion
void vmufb_present(const vmufb_t *fb, maple_device_t *dev) {
    vmu_draw_lcd_rotated(dev, fb->data);
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+	for (i = 0; i < h; i++) {
+		bits = extract_bits((const uint8_t *)data, i * w, w);
+		insert_bits((uint8_t *)fb->data, (y + i) * 48 + x, w, bits);
+	}
+}
+
+void vmufb_clear(vmufb_t *fb)
+{
+	memset(fb->data, 0, sizeof(fb->data));
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:82]:
```suggestion
void vmufb_clear(vmufb_t *fb) {
    memset(fb->data, 0, sizeof(fb->data));
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+void vmufb_clear_area(vmufb_t *fb,
+		      unsigned int x, unsigned int y,
+		      unsigned int w, unsigned int h)
+{
+	uint32_t tmp[48] = {};
+
+	vmufb_paint_area(fb, x, y, w, h, (const char *) tmp);
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:91]:
```suggestion
void vmufb_clear_area(vmufb_t *fb,
                      unsigned int x, unsigned int y,
                      unsigned int w, unsigned int h) {
    uint32_t tmp[48] = {};

    vmufb_paint_area(fb, x, y, w, h, (const char *) tmp);
}
```

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+			/* We ran out of space */
+			break;
+		}
+
+		vmufb_paint_area(fb, x, y, font->w, font->h,
+				 &font->fontdata[*str * font->stride]);
+
+		x += font->w;
+	}
+}
```
Comment (ljsebald) [inline kernel/arch/dreamcast/util/vmu_fb.c:132]:
```suggestion
void vmufb_print_string_into(vmufb_t *fb,
                             const vmufb_font_t *font,
                             unsigned int x, unsigned int y,
                             unsigned int w, unsigned int h,
                             unsigned int line_spacing,
                             const char *str) {
    unsigned int xorig = x, yorig = y;

    for(; *str; str++) {
        switch(*str) {
            case '\n':
                x = xorig;
                y += line_spacing + font->h;
               continue;

            default:
                break;
        }

        if(x > xorig + w - font->w) {
            x = xorig;
            y += line_spacing + font->h;
        }

        if(y > yorig + h - font->h) {
            /* We ran out of space */
            break;
        }

        vmufb_paint_area(fb, x, y, font->w, font->h,
                         &font->fontdata[*str * font->stride]);

        x += font->w;
    }
}
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 17]
```diff
@@ -0,0 +1,122 @@
+/* KallistiOS ##version##
+
+   dc/vmu_fb.h
+   Copyright (C) 2023 Paul Cercueil
+
+*/
+
+#include <kos.h>
+
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:9]: There should also be a Doxygen comment here explaining briefly what this header file contains.

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+/** \brief  VMU framebuffer.
+ *
+ *  This object can be used with the API below to compose a 48x32 image that can
+ *  then be displayed on the VMUs connected to the system.
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:14]:
```suggestion
/** \brief  VMU framebuffer.

    This object can be used with the API below to compose a 48x32 image that can
    then be displayed on the VMUs connected to the system.

    \headerfile dc/vmu_fb.h
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+typedef struct {
+	uint32 data[48];
+} vmufb_t;
+
+/** \brief  VMU framebuffer font meta-data. */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:19]:
```suggestion
/** \brief  VMU framebuffer font meta-data.

    \headerfile dc/vmu_fb.h
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+ *  \param  fb              A pointer to the vmufb_t to paint to.
+ *  \param  x               The horizontal position of the top-left corner of
+ *                          the drawing area, in pixels
+ *  \param  y               The vertical position of the top-left corner of the
+ *                          drawing area, in pixels
+ *  \param  w               The width of the drawing area, in pixels
+ *  \param  h               The height of the drawing area, in pixels
+ *  \param  data            A pointer to the pixel data that will be painted
+ *                          into the drawing area.
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:41]:
```suggestion
/** \brief  Render into the VMU framebuffer

    This function paints the provided pixel data into the VMU framebuffer,
    into the rectangle provided by the x, y, w and h values.

    \param  fb              A pointer to the vmufb_t to paint to.
    \param  x               The horizontal position of the top-left corner of
                            the drawing area, in pixels
    \param  y               The vertical position of the top-left corner of the
                            drawing area, in pixels
    \param  w               The width of the drawing area, in pixels
    \param  h               The height of the drawing area, in pixels
    \param  data            A pointer to the pixel data that will be painted
                            into the drawing area.
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+ * x, y, w and h values.
+ *
+ *  \param  fb              A pointer to the vmufb_t to paint to.
+ *  \param  x               The horizontal position of the top-left corner of
+ *                          the drawing area, in pixels
+ *  \param  y               The vertical position of the top-left corner of the
+ *                          drawing area, in pixels
+ *  \param  w               The width of the drawing area, in pixels
+ *  \param  h               The height of the drawing area, in pixels
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:59]:
```suggestion
/** \brief  Clear a specific area of the VMU framebuffer

    This function clears the area of the VMU framebuffer designated by the
    x, y, w and h values.

    \param  fb              A pointer to the vmufb_t to paint to.
    \param  x               The horizontal position of the top-left corner of
                            the drawing area, in pixels
    \param  y               The vertical position of the top-left corner of the
                            drawing area, in pixels
    \param  w               The width of the drawing area, in pixels
    \param  h               The height of the drawing area, in pixels
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+void vmufb_clear_area(vmufb_t *fb,
+		      unsigned int x, unsigned int y,
+		      unsigned int w, unsigned int h);
+
+/** \brief  Clear the VMU framebuffer
+ *
+ * This function will clear the whole VMU framebuffer.
+ *
+ *  \param  fb              A pointer to the vmufb_t to paint to.
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:69]:
```suggestion
/** \brief  Clear the VMU framebuffer

    This function clears the whole VMU framebuffer.

    \param  fb              A pointer to the vmufb_t to paint to.
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+void vmufb_clear(vmufb_t *fb);
+
+/** \brief  Present the VMU framebuffer to a VMU
+ *
+ * This function will present the previously rendered VMU framebuffer to the
+ * VMU identified by the dev argument.
+ *
+ *  \param  fb              A pointer to the vmufb_t to paint to.
+ *  \param  dev             The maple device of the VMU to present to
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:79]:
```suggestion
/** \brief  Present the VMU framebuffer to a VMU

    This function presents the previously rendered VMU framebuffer to the
    VMU identified by the dev argument.

    \param  fb              A pointer to the vmufb_t to paint to.
    \param  dev             The maple device of the VMU to present to
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+ *  \param  x               The horizontal position of the top-left corner of
+ *                          the drawing area, in pixels
+ *  \param  y               The vertical position of the top-left corner of the
+ *                          drawing area, in pixels
+ *  \param  w               The width of the drawing area, in pixels
+ *  \param  h               The height of the drawing area, in pixels
+ *  \param  line_spacing    Specify the number of empty lines that should
+ *                          separate two lines of text
+ *  \param  str             The text to render
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:99]:
```suggestion
/** \brief  Render a string into the VMU framebuffer

    This function uses the provided font to render text into the VMU
    framebuffer.

    \param  fb              A pointer to the vmufb_t to paint to.
    \param  font            A pointer to the vmufb_font_t that will be used for
                            painting the text
    \param  x               The horizontal position of the top-left corner of
                            the drawing area, in pixels
    \param  y               The vertical position of the top-left corner of the
                            drawing area, in pixels
    \param  w               The width of the drawing area, in pixels
    \param  h               The height of the drawing area, in pixels
    \param  line_spacing    Specify the number of empty lines that should
                            separate two lines of text
    \param  str             The text to render
 */
```

File: kernel/arch/dreamcast/include/dc/vmu_fb.h [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+/** \brief  Render a string into the VMU framebuffer
+ *
+ * Simplified version of vmufb_print_string_into. This is the same as calling
+ * vmufb_print_string_into with x=0, y=0, w=48, h=32, line_spacing=0.
+ *
+ *  \param  fb              A pointer to the vmufb_t to paint to.
+ *  \param  font            A pointer to the vmufb_font_t that will be used for
+ *                          painting the text
+ *  \param  str             The text to render
+ */
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/vmu_fb.h:116]:
```suggestion
/** \brief  Render a string into the VMU framebuffer

    Simplified version of vmufb_print_string_into(). This is the same as calling
    vmufb_print_string_into with x=0, y=0, w=48, h=32, line_spacing=0.

    \param  fb              A pointer to the vmufb_t to paint to.
    \param  font            A pointer to the vmufb_font_t that will be used for
                            painting the text
    \param  str             The text to render
 */
```

**Review by ljsebald — CHANGES_REQUESTED** [review]

File: kernel/arch/dreamcast/util/vmu_fb.c (hunk 8 above)
Comment (pcercuei) [reply kernel/arch/dreamcast/util/vmu_fb.c]: I'm too used to SPDX headers, I just forgot to remove it.

**Comment by pcercuei** [comment]
@ljsebald should be good now.

**Comment by gyrovorbis** [comment]
Dude, holy shit. I just got a chance to look at it, finally... I'm still trying to follow your bit manip craziness (lmao), but this API is incredible. This is like everything anyone could ever want for dynamic rendering on a VMU, then you just hooked right into the existing low-level drawing API... and even included font rendering.

I had been planning on doing something like this, and I thought I was going to be ambitious, but this is better than what I had had in mind. Wow.

File: kernel/arch/dreamcast/util/vmu_fb.c [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+
+void vmufb_clear_area(vmufb_t *fb,
+                      unsigned int x, unsigned int y,
+                      unsigned int w, unsigned int h) {
+    uint32_t tmp[48] = {};
+
+    vmufb_paint_area(fb, x, y, w, h, (const char *) tmp);
+}
+
+void vmufb_present(const vmufb_t *fb, maple_device_t *dev) {
```
Comment (gyrovorbis) [inline kernel/arch/dreamcast/util/vmu_fb.c:101]: Oops. I screwed up and over-simplified it. You have to look at the connection direction of BOTH the peripheral AND the controller to know how to orient... VMU is always 0 (or connector facing up) which is the same as regular controllers, but something like the lightguns return 1 (connector facing down). I think this should generalize it:
`void vmufb_present(const vmufb_t *fb, maple_device_t *dev) {
    maple_device_t* cont = maple_enum_dev(dev->port, 0);

    /* Check for controller containing VMU (should always be same port, unit 0) */
    if(cont && (cont->info.functions & MAPLE_FUNC_CONTROLLER)) {
        /* If the VMU connector and controller connector face opposite directions, 
           no flipping necessary (VMU in a lightgun). */
        if(cont->info.connector_direction != dev->info.connector_direction) {
            vmu_draw_lcd(dev, fb->data);
            return;
        }
    }
    /* If we somehow found no corresponding controller, or connectors face the same direction,
       we must rotate the image 180 degreees (VMU in a standard controller). */
    vmu_draw_lcd_rotated(dev, fb->data);
}`

**Review by ljsebald — APPROVED** [review]

File: kernel/arch/dreamcast/util/vmu_fb.c (hunk 26 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/util/vmu_fb.c:101]: Yep, no problem. I'm already in the VMU stuff adding date/time and button support. We're good!

**Pull request merged**